### PR TITLE
statestore: add migration capabilities

### PR DESCRIPTION
**Pull request by acud**
This PR adds migration capabilities to our leveldb statestore implementation. Implementation was adapted from the localstore migration code

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -74,6 +74,7 @@ require (
 	google.golang.org/protobuf v1.25.0 // indirect
 	gopkg.in/check.v1 v1.0.0-20200227125254-8fa46927fb4f // indirect
 	gopkg.in/ini.v1 v1.57.0 // indirect
+	gopkg.in/src-d/go-log.v1 v1.0.1
```
Comment (ralph-pichler) [inline go.mod:77]: why does this appear here again?

Comment (ralph-pichler) [reply go.mod:77]: please rebase and go mod tidy if necessary. this branch still contains the bad logger stuff from the deploy pr.

Comment (acud) [reply go.mod]: I reverted this change but I actually don't see this as a problem. Dunno where its coming from but the build is fine

File: pkg/statestore/leveldb/migration.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		if err != nil {
+			return err
+		}
+		s.logger.Infof("statestore: successfully ran migration: id %d current schema: %s", i, schemaName)
+	}
+	return nil
+}
+
+// getMigrations returns an ordered list of migrations that need be executed
+// with no errors in order to bring the localstore to the most up-to-date
```
Comment (ralph-pichler) [inline pkg/statestore/leveldb/migration.go:79]: comment says localstore

File: pkg/statestore/leveldb/migration_test.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
+// Copyright 2019 The Swarm Authors
```
Comment (ralph-pichler) [inline pkg/statestore/leveldb/migration_test.go:1]: the copyright headers here are all different than what we have in bee. also wrong license.

File: pkg/statestore/leveldb/leveldb.go [hunk 4]
```diff
@@ -17,16 +17,16 @@ import (
 	"github.com/syndtr/goleveldb/leveldb/util"
 )
 
-var _ storage.StateStorer = (*store)(nil)
+var _ storage.StateStorer = (*Store)(nil)
 
 // Store uses LevelDB to store values.
-type store struct {
+type Store struct {
```
Comment (ralph-pichler) [inline pkg/statestore/leveldb/leveldb.go:23]: why the export?

File: pkg/statestore/leveldb/migration_test.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+	dir, err := ioutil.TempDir("", "statestore-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	logger := logging.New(ioutil.Discard, 0)
+
+	// start the fresh localstore with the sanctuary schema name
```
Comment (ralph-pichler) [inline pkg/statestore/leveldb/migration_test.go:59]: comment says localstore (same in several other places)

File: pkg/statestore/leveldb/leveldb.go (hunk 4 above)
Comment (acud) [reply pkg/statestore/leveldb/leveldb.go:23]: because `leveldb_test`

Comment (janos) [reply pkg/statestore/leveldb/leveldb.go:23]: Maybe to export it in export_test.go? Or just to export there GetSchemaName and keep it unexported in the package.

File: pkg/statestore/leveldb/leveldb_test.go [hunk 6]
```diff
@@ -48,3 +48,33 @@ func TestPersistentStateStore(t *testing.T) {
 		return store
 	})
 }
+
+func TestGetSchemaName(t *testing.T) {
+	dir, err := ioutil.TempDir("", "statestore_test")
```
Comment (janos) [inline pkg/statestore/leveldb/leveldb_test.go:53]: I would suggest to use t.TempDir for convenience. Not just here but in all new code in this PR (6 instances).

**Review by janos — COMMENTED** [review]
I think that it is better to try to keep the `store` type unexported.

File: pkg/statestore/leveldb/leveldb.go (hunk 4 above)
Comment (acud) [reply pkg/statestore/leveldb/leveldb.go:23]: I tried it now again (second attempt). The problem is that the unit test needs to verify the schema name is the correct one, but the statestore constructor returns an interface. The unit test is not a black-box test as a result. This is why it was not a black-box test in localstore either (probably the only places in the codebase where this is necessary). The other idea I have is to unexport this, and just have a helper in the test that loads a new leveldb, and just tries to load the specified key out of the instance, and verify that it matches the latest schema version

**Review by esadakar — APPROVED** [review]

File: pkg/statestore/leveldb/leveldb.go (hunk 4 above)
Comment (acud) [reply pkg/statestore/leveldb/leveldb.go]: I didn't think of that :) Added a new commit, let me know if this is what you've meant

**Pull request merged**